d2/src/main/java/com/linkedin/d2/balancer/strategies/relative/ClientSelectorImpl.java

```diff
@@ -0,0 +1,158 @@
+                                        Ring<URI> ring,
+                                        Map<URI, TrackerClient> trackerClients)
+  {
+    TrackerClient trackerClient;
+
+    URI targetHostUri = KeyMapper.TargetHostHints.getRequestContextTargetHost(requestContext);
+
+    if (targetHostUri != null)
+    {
+      trackerClient = getTrackerClientFromTarget(targetHostUri, requestContext, trackerClients);
+    }
+    else
+    {
+      trackerClient = getTrackerClientFromRing(request, requestContext, ring, trackerClients);
+    }
+
+    return trackerClient;
+  }
+
+  private void addToExcludedHosts(TrackerClient trackerClient, RequestContext requestContext)
+  {
+    if (trackerClient != null)
+    {
+      LoadBalancerStrategy.ExcludedHostHints.addRequestContextExcludedHost(requestContext, trackerClient.getUri());
+    }
+  }
+
+  private TrackerClient getTrackerClientFromTarget(URI targetHostUri, RequestContext requestContext, Map<URI, TrackerClient> trackerClients)
+  {
+    TrackerClient trackerClient = trackerClients.get(targetHostUri);
+
+    if (trackerClient == null)
+    {
+      LOG.warn("No client found for ", targetHostUri, ". Target host specified is no longer part of cluster");
+    }
+    else
+    {
+      // if this flag is set to be true, that means affinity routing is preferred but backup requests are still acceptable
+      Boolean otherHostAcceptable = KeyMapper.TargetHostHints.getRequestContextOtherHostAcceptable(requestContext);
+      if (otherHostAcceptable != null && otherHostAcceptable)
+      {
+        addToExcludedHosts(trackerClient, requestContext);
+      }
+    }
+    return trackerClient;
+  }
+
+  private TrackerClient getTrackerClientFromRing(Request request,
+                                                 RequestContext requestContext,
+                                                 Ring<URI> ring,
+                                                 Map<URI, TrackerClient> trackerClients)
+  {
+    Set<URI> excludedUris = LoadBalancerStrategy.ExcludedHostHints.getRequestContextExcludedHosts(requestContext) == null
+        ? new HashSet<>()
+        : LoadBalancerStrategy.ExcludedHostHints.getRequestContextExcludedHosts(requestContext);
+    int hashCode = _requestHashFunction.hash(request);
+    URI uri = ring.get(hashCode);
+
+    TrackerClient trackerClient = trackerClients.get(uri);
+
+    if (trackerClient == null || excludedUris.contains(uri))
+    {
+      // Find next available URI.
+      Iterator<URI> ringIterator = ring.getIterator(hashCode);
+
+      while (ringIterator.hasNext())
+      {
+        uri = ringIterator.next();
+        trackerClient = trackerClients.get(uri);
+
+        if (trackerClient != null && !excludedUris.contains(uri))
+        {
+          break;
+        }
+      }
+      trackerClient = null;
+    }
+
+    if (trackerClient == null)
+    {
+      // Pick one from the tracker clients passed from the request if the ring is completely out of date
```

Comment:
This may happen a little bit more often compared with the existing V3 implementation. Because we do more async partition state update, if a request comes in and the cluster uris just changed, because we use executor to trigger the update, we may end up with a completely out-of-date ring for this request if the cluster uri set is completely new. The possibility is higher than the old implementation, but it is still very rare.

Reply:
It's hardly to have a completely new list of URIs so agree it's rare.